### ISSUE TITLE
Add hook which is executed only once when starting container for the first time

### ISSUE THREAD

**Issue by tmotyl** [drud/ddev#1115]
**Is your feature request related to a problem? Please describe.**
Often the post-start hook is used to install additional software (apt update && apt install...).
This can take some time, it's quite annoying that you have to wait for ddev to run the commands every time you boot the container.

It would be enough to run the commands just once after the container has been created, instead of running them on every start.
So if somebody needs to run the hook again, he need to `ddev rm` and `ddev start`.

The current workaround I'm using is to comment the hooks once they were executed once, but it's really ugly and error prone.

I propose a new hook called "post-start-once":
It would look like:
```
hooks:
  post-start-once:
    - exec: "apt-get update"
    - exec: "apt-get install my package"
```

**Comment by rfay** [comment]
I'd be great with this, but don't know how to do it.   I haven't ever figured out when docker-compose will rebuild and when it won't. That's my biggest problem. At this time, ddev is not involved in that choice, it just does a docker-compose up. And it doesn't seem completely predictable. I have thought we should move to *always* rebuilding (`docker-compose up --force-recreate`), which would give very predictable results. Maybe that's the ticket. But then you wouldn't need this hook  :)

**Comment by rfay** [comment]
I think we might be able to solve this with this simple technique:
* After hook execution, `touch /var/tmp/ddev_startup_complete.txt`
* Only execute post-start hooks if that file does *not* exist.

**Comment by tmotyl** [comment]
Yay!

**Comment by rfay** [comment]
Hope you like it :)

**Comment by rfay** [comment]
Actually, would you mind giving a quick test? Artifacts are in https://circleci.com/gh/drud/ddev/6762#artifacts/containers/0 - I meant to ask you for a test run of this before it got pulled @tmotyl . Note that it changes the behavior slightly - post-start now happens only when web container is rebuilt. You had originally asked for a different hook, but it seemed like this is what people would actually want.

**Comment by fnagel** [comment]
@rfay The option to only run a hook on the first start (e.g. docker-compose up) has been removed, right? Looks like the changes in 9794512f8c8f5bf441ecd76ea0bdc9a6fda4cb78 are no longer there, but been removed here: 6ba9fd61c8f1984b165e928523de282b6e412fcf

**Comment by jonaseberle** [comment]
`post-start` hooks run on `ddev start` (and I like it that way).

What's your use case @fnagel? 

Maybe you want to rather do it in a `Dockerfile` at build time or like this in a `post-start` hook?
```
stat /tmp/has_run &> /dev/null || my_command && touch /tmp/has_run
```

**Comment by fnagel** [comment]
@jonaseberle Thanks for the quick response! So it has been removed, right?

My use case is that I want to add some stuff as a `config.local.yaml` in order to make it work as the only Windows user of a project group (all other use Apple). In detail: I disabled the creation of the mounts and then copy files into the container using docker cp. 

It should be possible using a custom Dockerfile. Is that something I can change for me only (and not for all other members for the project).

I will test the idea with the stat command. Quick and dirty is good for now ;-)

Thanks a lot!

**Comment by fnagel** [comment]
ps: another proper use case would be an initial `composer install` which should be done once on create of the container.

**Comment by jonaseberle** [comment]
Maybe it's worth discussing your cross-OS team concerns on Slack or as a new issue. I think it might trigger some interesting ideas. I am not so much into it. We just use defaults for all this filesystem stuff. MacOS users have `mutagen_enabled: true` in their [global ddev config](https://ddev.readthedocs.io/en/stable/users/extend/config_yaml/#global_configyaml-options).

Regarding `composer install`: We have it run on `ddev start` (together with asset building and cache flushing) and it is a blessing IMHO. Whenever someone checks out another code state a `ddev branch` is all that is needed to "build" the instance properly for the current code state.
I really like that I can tell co-workers and fellow TYPO3 team members: If you experience problems, a `ddev start` will most likely fix them.
(In some projects I have moved to `make` with a `Makefile`: That can be built so that it only runs build tasks that are actually needed due to changed files/directories and works fine with composer -> container startup time is greatly reduced)

**Comment by rfay** [comment]
For installing extra software use webimage_extra_packages if a standard debian package or a .ddev/web-build/Dockerfile for more complex things. This is far better than the hook you propose because it only happens the first time you start a new project.

https://ddev.readthedocs.io/en/latest/users/extend/customizing-images/
